File: backend/schemas/scene.py
from __future__ import annotations

# JSON scene: boundary, template ref, layers, batches, background, tool.

from typing import Literal

from pydantic import BaseModel, ConfigDict, Field, field_validator
from hideout_core.config.editor_assets import (
    asset_specs_by_key,
    drawing_asset_keys,
    legacy_asset_aliases,
    load_editor_assets_manifest,
)

SCENE_VERSION = 2
MIN_SCENE_COORD = -1_000_000
MAX_SCENE_COORD = 1_000_000
MIN_SCENE_ROTATION = -360_000
MAX_SCENE_ROTATION = 360_000
MAX_TEMPLATE_HASH = 4_294_967_295

PAINT_LAYER_KINDS = ("default", "decorations", "palette", "user")
DRAW_STYLES = ("line", "brush", "object")
TOOL_MODE_VARIANTS = ("eraser", "select", "line", "fill")
FILL_MODES = (
    "four_way",
    "eight_way_free",
    "eight_way_corner_safe",
    "orthogonal_first",
    "radius_limited",
    "narrow_passage_block",
    "weighted",
)
FILL_CONNECTIVITY = (4, 8)
FILL_WALLS_SCOPE = ("all_layers", "active_layer")
TOOL_VARIANTS = (*drawing_asset_keys(), *TOOL_MODE_VARIANTS)
# Runtime import seam: persisted/uploaded scenes may still arrive with a legacy
# alias such as `rope`, even though the editor now stores manifest-backed asset
# keys in the canonical tool contract.
LEGACY_TOOL_VARIANTS = tuple(legacy_asset_aliases().keys())
ACCEPTED_TOOL_VARIANTS = (*TOOL_VARIANTS, *LEGACY_TOOL_VARIANTS)
DEFAULT_TOOL_VARIANT = str(load_editor_assets_manifest()["toolFvAssetKey"])
DEFAULT_TOOL_FV = int(asset_specs_by_key()[DEFAULT_TOOL_VARIANT]["defaultFv"])


def build_editor_scene_contract_artifact() -> dict[str, object]:
    return {
        "sceneVersion": SCENE_VERSION,
        "paintLayerKinds": list(PAINT_LAYER_KINDS),
        "drawStyles": list(DRAW_STYLES),
        "toolVariants": list(TOOL_VARIANTS),
        "legacyToolVariants": list(LEGACY_TOOL_VARIANTS),
        "acceptedToolVariants": list(ACCEPTED_TOOL_VARIANTS),
        "fillModes": list(FILL_MODES),
        "fillConnectivity": list(FILL_CONNECTIVITY),
        "fillWallsScope": list(FILL_WALLS_SCOPE),
    }


class XY(BaseModel):
    x: int = Field(ge=MIN_SCENE_COORD, le=MAX_SCENE_COORD)
    y: int = Field(ge=MIN_SCENE_COORD, le=MAX_SCENE_COORD)


class XYZRPlacement(BaseModel):
    x: int = Field(ge=MIN_SCENE_COORD, le=MAX_SCENE_COORD)
    y: int = Field(ge=MIN_SCENE_COORD, le=MAX_SCENE_COORD)
    r: int = Field(ge=MIN_SCENE_ROTATION, le=MAX_SCENE_ROTATION)


class BoundaryModel(BaseModel):
    points: list[XY] = Field(min_length=3)


class PaintedBatchModel(BaseModel):
    template_name_ru: str
    template_hash: int = Field(ge=0, le=MAX_TEMPLATE_HASH)
    placements: list[XYZRPlacement] = Field(default_factory=list)
    facet_fv: int | None = Field(
        default=None,
        ge=0,
        description="Fv в экспорте .hideout; если None — берётся tool.fv сцены.",
    )
    line_stroke: bool | None = Field(
        default=None,
        description="Штрих «Линия»; r хранится для экспорта .hideout (frontend может применять preview-коррекцию legacy-спрайтов).",
    )


class PaintLayerModel(BaseModel):
    kind: Literal["default", "decorations", "palette", "user"] | None = None
    title: str | None = None
    visible: bool = True
    locked: bool = False
    batches: list[PaintedBatchModel] = Field(default_factory=list)


class ToolModel(BaseModel):
    # Tool settings are intentionally forward-compatible: frontend may persist
    # tool-specific options unknown to this backend version.
    model_config = ConfigDict(extra="allow")

    variant: str = DEFAULT_TOOL_VARIANT
    draw_style: Literal["line", "brush", "object"] = "object"
    spacing: float = 3.0
    margin: float = 3.0
    fv: int = DEFAULT_TOOL_FV
    brush_width_view: float = 24.0
    fill_step_world: float | None = Field(
        default=None,
        description="Заливка: шаг обхода world-сетки и расстояние между соседними объектами (>=1).",
    )
    fill_max_placements: int | None = Field(
        default=None,
        description="Заливка: максимум добавляемых объектов за один запуск.",
    )
    fill_margin_world: float | None = Field(
        default=None,
        description=(
            "Заливка: отступ от края зоны (мир). Может быть отрицательным — "
            "тогда заливка заходит за границу. UI отображает значение со "
            "смещением +4 (дефолт −4 показан как 0)."
        ),
    )
    fill_mode: Literal[
        "four_way",
        "eight_way_free",
        "eight_way_corner_safe",
        "orthogonal_first",
        "radius_limited",
        "narrow_passage_block",
        "weighted",
    ] | None = Field(
        default=None,
        description="Заливка: режим обхода области.",
    )
    fill_mode_params: dict[str, float] | None = Field(
        default=None,
        description=(
            "Параметры режима заливки: radius_world, min_passage_width_world, "
            "cardinal_cost, diagonal_cost."
        ),
    )
    fill_connectivity: Literal[4, 8] | None = Field(
        default=None,
        description="Заливка: связность flood-fill (4-way или 8-way).",
    )
    fill_walls_scope: Literal["all_layers", "active_layer"] | None = Field(
        default=None,
        description="Заливка: какие слои считать стенами (все или только активный).",
    )
    eraser_targets: dict[str, bool] | None = Field(
        default=None,
        description=(
            "Фронт: для стерки — какие asset key удалять "
            "(включая faridun_ropes4/faridun_ropes1/moss/sand и др.)."
        ),
    )

    @field_validator("variant")
    @classmethod
    def validate_variant(cls, v: str) -> str:
        vv = str(v).strip()
        if vv not in ACCEPTED_TOOL_VARIANTS:
            raise ValueError(
                "Input should be one of: "
                + ", ".join(repr(x) for x in ACCEPTED_TOOL_VARIANTS),
            )
        return vv


class BackgroundModel(BaseModel):
    path: str = ""
    opacity: float = 0.55
    scale: float = 1.0
    rotation_deg: float = 0.0
    crop_left_pct: float = 0.0
    crop_top_pct: float = 0.0
    crop_right_pct: float = 100.0
    crop_bottom_pct: float = 100.0
    offset_x: float = 0.0
    offset_y: float = 0.0
    width_view_base: float | None = None
    locked: bool = False
    lock_anchor_view_x: float | None = None
    lock_anchor_view_y: float | None = None
    asset_id: str | None = None


class TemplateRefModel(BaseModel):
    # Server-side .hideout template id; empty means draft (export unavailable).

    template_id: str = ""


class UiModel(BaseModel):
    drawing_enabled: bool = True
    show_template_dots: bool = True
    placement_preview_scale: float = Field(
        default=1.0,
        ge=0.2,
        le=8.0,
        description="Множитель радиуса маркеров размещения в превью редактора.",
    )


class SceneModel(BaseModel):
    model_config = ConfigDict(extra="ignore")

    scene_version: Literal[2] = SCENE_VERSION
    camera_deg: float = Field(default=45.0, ge=-180, le=180)
    boundary: BoundaryModel
    template: TemplateRefModel = Field(default_factory=TemplateRefModel)
    layers: list[PaintLayerModel] = Field(min_length=1)
    tool: ToolModel = Field(default_factory=ToolModel)
    background: BackgroundModel = Field(default_factory=BackgroundModel)
    ui: UiModel | None = None
    template_dots_cache: list[XY] | None = Field(
        default=None,
        description="Кэш точек шаблона при сохранении сцены в файл (не в .hideout).",
    )
    hideout_map_display_name: str | None = Field(
        default=None,
        description="Имя карты hideout (из SQLite), к которой привязана текущая работа с границей.",
    )
    lineage_base_display_name: str | None = Field(
        default=None,
        description=(
            "Базовая карта по цепочке fork (встроенная или пользовательская); "
            "для экспорта .hideout — выбор hideout_name в шапке файла."
        ),
    )


class TemplateLoadResponse(BaseModel):
    template_id: str
    hideout_name: str | None = None
    hideout_hash: int | None = None
    doodads_kept_count: int
    dots: list[XY]
    default_layer_batches: list[PaintedBatchModel] = Field(
        default_factory=list,
        description="Плейсменты из doodads для слоя 0 редактора (по одному батчу на doodad).",
    )
    decorations_layer_batches: list[PaintedBatchModel] = Field(
        default_factory=list,
        description="Украшения без «точной» палитры — слой 1 при разбиении или все украшения.",
    )
    decorations_palette_layer_batches: list[PaintedBatchModel] = Field(
        default_factory=list,
        description="Верёвка/мох/песок как в палитре (hash+fv) — слой 2 при импорте.",
    )

File: frontend/src/types/scene.ts
// JSON scene type for the hideout-editor web app.
import type { EditorAssetKey } from '../shared/generated/editorAssets';
import {
    DRAW_STYLES as GENERATED_DRAW_STYLES,
    FILL_CONNECTIVITY as GENERATED_FILL_CONNECTIVITY,
    FILL_MODES as GENERATED_FILL_MODES,
    FILL_WALLS_SCOPE as GENERATED_FILL_WALLS_SCOPE,
    PAINT_LAYER_KINDS as GENERATED_PAINT_LAYER_KINDS,
    SCENE_VERSION as GENERATED_SCENE_VERSION,
    TOOL_VARIANTS as GENERATED_TOOL_VARIANTS,
} from '../shared/generated/editorSceneContract';

export type PaintVariant = (typeof GENERATED_TOOL_VARIANTS)[number];
// Runtime tool list sourced from the generated backend contract artifact.
export const TOOL_VARIANTS = [...GENERATED_TOOL_VARIANTS] as PaintVariant[];
export type AssetKey = EditorAssetKey;
export type DrawStyle = (typeof GENERATED_DRAW_STYLES)[number];
export type FillConnectivity = (typeof GENERATED_FILL_CONNECTIVITY)[number];
export type FillWallsScope = (typeof GENERATED_FILL_WALLS_SCOPE)[number];
export type FillMode = (typeof GENERATED_FILL_MODES)[number];
export const FILL_MODES = [...GENERATED_FILL_MODES] as FillMode[];
export const SCENE_VERSION = GENERATED_SCENE_VERSION;
export type FillModeParams = {
    radius_world?: number;
    min_passage_width_world?: number;
    cardinal_cost?: number;
    diagonal_cost?: number;
};

export interface XY {
    x: number;
    y: number;
}

export interface XYZRPlacement {
    x: number;
    y: number;
    r: number;
}

export interface Boundary {
    points: XY[];
}

export interface PaintedBatch {
    template_name_ru: string;
    template_hash: number;
    placements: XYZRPlacement[];
    // When set, the .hideout export writes this fv on the batch doodads.
    facet_fv?: number | null;
    // Stored r value as in .hideout; faridun_ropes4 also receives a preview-time visual correction.
    line_stroke?: boolean;
}

export type PaintLayerKind = (typeof GENERATED_PAINT_LAYER_KINDS)[number];

export interface PaintLayer {
    // Stable semantic layer kind used for locale-aware UI labels.
    kind?: PaintLayerKind;
    // Optional user-defined title. System layers should prefer `kind`.
    title?: string;
    visible: boolean;
    locked: boolean;
    batches: PaintedBatch[];
}

export interface Tool {
    variant: PaintVariant;
    draw_style: DrawStyle;
    // Last asset chosen in single-asset mode.
    asset_key?: AssetKey;
    // Asset chosen for the Line tool. Does not affect Fill.
    line_asset_key?: AssetKey;
    // Asset chosen for the Fill tool. Does not affect Line.
    fill_asset_key?: AssetKey;
    spacing: number;
    margin: number;
    fv: number;
    brush_width_view: number;
    // Fill: world-grid traversal step and distance between neighbouring placements (>= 1).
    fill_step_world?: number;
    // Fill: cap on placements added in a single run.
    fill_max_placements?: number;
    // Fill: zone-edge margin (world units). Negative values allow fill slightly
    // outside the boundary. UI displays the value with a +4 offset (default
    // -4 maps to 0 on screen).
    fill_margin_world?: number;
    // Fill: main flood-fill traversal mode.
    fill_mode?: FillMode;
    // Extra fill-mode parameters (radius, passage width, step weights).
    fill_mode_params?: FillModeParams;
    // Fill: flood-fill connectivity (4-way or 8-way).
    fill_connectivity?: FillConnectivity;
    // Fill: which layers are treated as walls.
    fill_walls_scope?: FillWallsScope;
    // When variant === 'eraser': which decoration kinds the eraser affects. Missing key means "all".
    eraser_targets?: Partial<Record<AssetKey, boolean>>;
}

export interface Background {
    path?: string;
    opacity?: number;
    scale?: number;
    rotation_deg?: number;
    crop_left_pct?: number;
    crop_top_pct?: number;
    crop_right_pct?: number;
    crop_bottom_pct?: number;
    offset_x?: number;
    offset_y?: number;
    width_view_base?: number | null;
    locked?: boolean;
    lock_anchor_view_x?: number | null;
    lock_anchor_view_y?: number | null;
    asset_id?: string | null;
}

export interface TemplateRef {
    template_id: string;
}

export interface UiState {
    drawing_enabled: boolean;
    show_template_dots: boolean;
    placement_preview_scale: number;
}

export interface Scene {
    scene_version: number;
    camera_deg: number;
    boundary: Boundary;
    template: TemplateRef;
    layers: PaintLayer[];
    tool: Tool;
    background: Background;
    ui?: UiState;
    template_dots_cache?: XY[] | null;
    hideout_map_display_name?: string | null;
    // Base map along the fork chain. Used for the .hideout export's
    // hideout_name (built-in or user-created base).
    lineage_base_display_name?: string | null;
}

export interface TemplateUploadResponse {
    template_id: string;
    hideout_name: string | null;
    hideout_hash: number | null;
    doodads_kept_count: number;
    dots: XY[];
    // Doodad placements for the "Default objects" layer (index 0).
    default_layer_batches?: PaintedBatch[];
    // Decorations without palette / non-standard fv: layer 1 when split, otherwise all decorations.
    decorations_layer_batches?: PaintedBatch[];
    // Palette decorations (rope, moss, sand identified by hash+fv): layer 2.
    decorations_palette_layer_batches?: PaintedBatch[];
}
